### ISSUE TITLE
Provisioning of VM extension 'vmssCSE' has timed out

### ISSUE THREAD

**Issue by vijaygos** [Azure/aks-engine#1860]
**What happened**:
VMSS status is set to 'failed' with error message - "Provisioning of VM extension 'vmssCSE' has timed out. Extension installation may be taking too long, or extension status could not be obtained." 

As a result of this, SLB does not allow for binding of service (type Load balancer) with a public IP resource. The service status is always Pending:

> NAME...................TYPE..................... CLUSTER-IP.......EXTERNAL-IP......PORT(S) ........................................AGE

> my-service-1---ClusterIP -------10.0.52.61-----`<none>` ---------8125/UDP -------------------57m

> my-service-2---LoadBalancer---10.0.165.41 ---**`<pending>`**----80:31330/TCP,443:30354/TCP ---46m

**What you expected to happen**:
No CSE errors and the service should bind to a given public IP resource with no errors.
**How to reproduce it (as minimally and precisely as possible)**:
No real steps. Happens at random when the cluster attempts to scale and add a new VM

**Anything else we need to know?**:
AKS engine version is 0.28.1
**Environment**:
- Kubernetes version (use `kubectl version`): 1.12.2
- Size of cluster (how many worker nodes are in the cluster?) 31
- General description of workloads in the cluster (e.g. HTTP microservices, Java app, Ruby on Rails, machine learning, etc.)
- Others: Azure US Government environment

While I am tempted to say this looks like a duplicate of #802, I would appreciate another look.

**Comment by chpspiir** [comment]
We are experiencing the exact same thing. Any updates on what could be wrong?

**Comment by rsingh612** [comment]
Also, Experiencing the same issue after updating ServicePrincipal credentials 
'Provisioning of VM extension 'vmssCSE' has timed out. Extension installation may be taking too long, or extension status could not be obtained.'

**Comment by sylus** [comment]
Also getting this issue today though no changes from cluster perspective. We were getting a rate limit issue earlier.

**Comment by jackfrancis** [comment]
Is there a common pattern in terms of cluster operations? Are these new cluster buildouts? Or a result of scale events on existing clusters?

**Comment by zachomedia** [comment]
Answering for @sylus - it's an existing cluster. We haven't made changes to the scale set ourselves - the last scale operation was last week. We found the issue when we had teams reporting that pods with disks weren't coming up. 

It now seems that it's unable to mount the disks because it can't unmount them from the old nodes.

We seemed to hit a subscription write limit earlier today - though I'm not sure if that's related to this issue or not (if it was retrying too often).

**Comment by MoergJ** [comment]
We experience the same issue. Existing cluster, no changes regarding scale sets, but pods cant mount data disks and keep hanging in 'Init'.

**Comment by lorenzo-biava** [comment]
Same here; it worked fine last evening and today no VM is able to attach any disks, displaying the `Provisioning of VM extension 'vmssCSE' has timed out.` failure state in the VMSS (the AKS cluster is in westeurope).

**Comment by sylus** [comment]
We filed an Azure support ticket and supposed to get a call back this morning will post back any info.

**Comment by rsingh612** [comment]
I was able to restart our VMSS nodes gracefully by running the following az cli command. Previously restarting the vmss node using the GUI also resulted in  `Provisioning of VM extension 'vmssCSE' has timed out.`

Command:
`az vmss update-instances -g {RESOUCE_GROUP} --name {VMSS_AGENTPOOL} --instance-id {INSTANCE_ID}`

And the Cluster status resolved, All hung up Disks freed themselves and reattach without an issue.

**Comment by jarobar435** [comment]
Unfortunately, in my case, the outcome of the above wasn't any different than the GUI triggered process:
```
$ az vmss update-instances -g {group} --name {vmss} --instance-id {id}
Deployment failed. Correlation ID: {id}. Provisioning of VM extension 'vmssCSE' has timed out. Extension installation may be taking too long, or extension status could not be obtained.
```

**Comment by sharkymcdongles** [comment]
We have the same issue except it affects both vmssCSE and RunCommandLinux extensions. We didn't enable or disable any extensions. I tried to do the update command but it never completes successfully failing with this:

```
cli.azure.cli.core.util : Deployment failed. Correlation ID: abeff543-0a78-4c26-aa0f-dc2d01b71222. Provisioning of VM extension 'RunCommandLinux' has timed out. Extension installation may be taking too long, or extension status could not be obtained.
Deployment failed. Correlation ID: abeff543-0a78-4c26-aa0f-dc2d01b71222. Provisioning of VM extension 'RunCommandLinux' has timed out. Extension installation may be taking too long, or extension status could not be obtained.
```
I  tried to reinstall the extension with force and it also failed. This is a critical bug and is affecting our production. How has this not been addressed? We don't need this extension. Nobody needs to run linux commands via a cli extension and this should not be there by default. It's a bad design period to force this on people.

**Comment by jackfrancis** [comment]
How many folks on this thread can correlate these failures with active Azure API rate limiting events?

@rsingh612, were you able to determine operational differences between your original failure events, and your successful (if manual) `az vmss update-instances` remediation?

**Comment by sharkymcdongles** [comment]
@jackfrancis we saw this:

` failed with failed to get azure instance id for node "aks-e16sv3-79416110-vmss00000w" (compute.VirtualMachineScaleSetVMsClient#Get: Failure responding to request: StatusCode=429 -- Original Error: autorest/azure: Service returned an error. Status=429 Code="OperationNotAllowed" Message="The server rejected the request because too many requests have been received for this subscription." Details=[{"code":"TooManyRequests","message":"{\"operationGroup\":\"GetVMScaleSetVM30Min\",\"startTime\":\"2019-10-18T08:32:36.6723441+00:00\",\"endTime\":\"2019-10-18T08:47:36.6723441+00:00\",\"allowedRequestCount\":2500,\"measuredRequestCount\":2843}","target":"GetVMScaleSetVM30Min"}] InnerError={"internalErrorCode":"TooManyRequestsReceived"})`

I assumed what happened was a vm extension autoupdated leading to the vmss enterring a failed state. While in a failed state disk attach detach operations fail but will keep retrying. My hypothesis was maybe because of the failed state the disks retried triggering the api limit, but maybe that is an incorrect hypothesis.

**Comment by jackfrancis** [comment]
@feiskyer @khenidak will cloudprovider backoff v2 help recover from these events?

**Comment by sylus** [comment]
We are now getting this on a completely different cluster. Are we 100% there was no roll out change? It seems suspicious with all the people here having same problem and over in:

https://github.com/Azure/AKS/issues/1278

Obviously something had to change, is Microsoft looking into this?

**Comment by sylus** [comment]
We were able to add a node and that process seems to have fixed the scaleset at least for now. ^_^

**Comment by devigned** [comment]
We are actively working on a resolution for this issue.

**Comment by sharkymcdongles** [comment]
Adding a node didn't work for us sadly. 

@devigned  Any idea on when you might expect a resolution to this and whether or not it will involve upgrading or migrating to a new aks / nodepool?

**Comment by devigned** [comment]
It seems like most of these issue are related to Azure API request limits being exceeded. It would be extremely helpful if everyone experiencing this issue could ping back with the following:
- Number of nodes in your cluster?
- Number of disks attached to those nodes?
- Is cluster auto scaler enabled?
- What operation did you run when this issue was triggered?
- What version of aks-engine?
- What version of K8s?

Thank you all. We will do our best to root cause this and get a fix out as soon as possible.

**Comment by sharkymcdongles** [comment]
Number of nodes in your cluster?
30

Number of disks attached to those nodes?
varies but we have 131 pvcs with 5 or so a nod
Is cluster auto scaler enabled?
no it is disabled
What operation did you run when this issue was triggered?
we didn't do anything. the cluster just stopped mounting pvcs and said there was an api limit. we then checked and saw the vmss was in a failed state due to vm extension RunCommandLinux and vmssCSE
What version of aks-engine?
aksEngineVersion : v0.39.2-aks

What version of K8s?
1.14.6

**Comment by devigned** [comment]
@sharkymcdongles give this a run please:
```
kubectl logs {{ your master named like: kube-controller-manager-k8s-master-XXXXXXX-0 }} \
-n kube-system | grep azure
```

**Comment by sharkymcdongles** [comment]
@devigned this is aks, I dont have access to the masters.

this is the affected vmss in azure:

`/subscriptions/63ffcd66-455e-4369-87e6-a4a55983f44e/resourceGroups/MC_prod-eu-mip-aks-rg_euprod_westeurope/providers/Microsoft.Compute/virtualMachineScaleSets/aks-e16sv3-79416110-vmss`

and aks: 

`/subscriptions/63ffcd66-455e-4369-87e6-a4a55983f44e/resourcegroups/prod-eu-mip-aks-rg/providers/Microsoft.ContainerService/managedClusters/euprod`

**Comment by CecileRobertMichon** [comment]
@sharkymcdongles for AKS cluster issues, please open a support ticket and/or open an issue at https://github.com/Azure/AKS/issues.

**Comment by sharkymcdongles** [comment]
There is an issue there already but here is the only place with engineers
replying lol. Also it still uses aks engine even on aks hence the name so
this bug is relevant.

**Comment by andyzhangx** [comment]
one workaround is update vmss instance that has issue manually:
```
az vmss update-instances -g <RESOURCE_GROUP_NAME> --name <VMSS_NAME> --instance-id <ID(number)>
```
Let me know whether that works.

**Comment by sharkymcdongles** [comment]
@andyzhangx that was the first thing I tried. It fails with an internal execution error. I then tried deallocating one of the bad nodes but now it won't start with an internal execution error as well.

**Comment by rsingh612** [comment]
@jackfrancis

operational differences
After failure ( service principal update)
* Pods were stuck in Terminating status
* Statefulset pods were unable to attach persistent volumes.
* Vmss resource in Azure portal status : Failure

After executing upgrade 
* Cluster and VMSS back to functional status

**Comment by abarqawi** [comment]
Please detach the disk explicitly , it should work

 az vmss disk detach -g MC_kube_xyz_abc  --name  aks-d8s-27745123-vmss  --instance-id 0  --lun 0

where 0 is vm 0 so it can be 0,1,2
 --name  aks-d8s-27745123-vmss : **vmss name**

**Comment by lorenzo-biava** [comment]
@abarqawi nice idea, but in my case still fails:
> Deployment failed. Provisioning of VM extension 'vmssCSE' has timed out. Extension installation may be taking too long, or extension status could not be obtained.

However, after a while (hours) the disk might get detached/attached successfully.

**Comment by andreygolev** [comment]
Joining to the club. Same issues with both vmssCE extension and stuck pods waiting for storage

**Comment by ahmedspiir** [comment]
Same here problem with vmssCE timing out

**Comment by acm-073** [comment]
Same problem here.

The workaround above helped in our case (run update-instances for each instance, run reimage for instances where update-instances returns an error).

However, after the procedure, the azure portal still shows the VMSS as "failed" with the error message saying "Provisioning of VM extension 'vmssCSE' has timed out."

**Comment by sylus** [comment]
And now our third cluster with VMSS is doing this in a different subscription after this happened to two other clusters. Definitely starting to get kind of frustrating lol.

**Comment by sylus** [comment]
Can we get some sort of statement from Azure into what is going on? I think there are enough people this issue is happening to warrant some sort of explanation? I am super concerned about this.

**Comment by jackfrancis** [comment]
Is anyone on this thread suffering these cluster symptoms where that cluster is the *only* cluster in a particular subscription? That would make reproducing the error state easier, thus accelerating root cause diagnosis.

**Comment by devigned** [comment]
@sylus, would you be up for jumping on a call and having a shared debugging session?

**Comment by sylus** [comment]
@devigned we have gotten our clusters back in a working state for now but super happy to have a chat on the problems we were having and provide any information that will assist you.

We are free from 2-4:30 EDT, or can find a time tomorrow if need :D

Thanks a bunch!

**Comment by jackfrancis** [comment]
@andreygolev what version of Kubernetes is the cluster (or clusters) running that are suffering rate throttling?

**Comment by sharkymcdongles** [comment]
Mine is 1.14.6

**Comment by lorenzo-biava** [comment]
Same here.

**Comment by jackfrancis** [comment]
@sharkymcdongles If you're still in a sticky spot, you may try the following to gain cloudprovider improvements w/ respect to Azure API calls:

For each master node:

- ssh into the vm
- `sudo vi /etc/kubernetes/manifests/kube-controller-manager.yaml`
- change the image reference to 1.14.8 (from 1.14.6 in your case)
  - e.g., look for `image: k8s.gcr.io/hyperkube-amd64:v1.14.6` and change to `image: k8s.gcr.io/hyperkube-amd64:v1.14.8`

There are several optimizations to VMSS-based Kubernetes resource reconciliations that are in 1.13.12, 1.14.8, 1.15.5, and 1.16.2. Those versions were only just added in an aks-engine release today:

https://github.com/Azure/aks-engine/releases/tag/v0.42.1

For folks who have a working `aks-engine upgrade` workflow in place, we recommend getting those versions @ v0.42.1 of aks-engine. For folks managing cluster in active throttling situations, we don't recommend executing an `aks-engine upgrade`, as that operation will itself require a non-throttled Azure API subscription in order to succeed. :(

An additional step-wise remediation is to apply a newer version of "exponential backoff" to the cloudprovider config by adding this line to `/etc/kubernetes/azure.json` (again, on all master vms):

```
$ sudo cat /etc/kubernetes/azure.json | grep cloudProviderBackoffMode
    "cloudProviderBackoffMode": "v2",
```

(The above output is from a test cluster running that config.)

The above configuration optimizes the number of calls that are stacked up and continually sent to the Azure APIs in HTTP 429 situations, helping most scenarios to back off more effectively, in order to make Azure API throttling symptoms go away sooner.

**Comment by devigned** [comment]
@sylus glad to hear your cluster is back in a working state! Thank you for the kind words and offer for additional info. We think we have narrowed down the issue and we are hoping https://github.com/Azure/aks-engine/issues/1860#issuecomment-544694755 will reduce the API throttling we think is behind these issues.

Furthermore, we have some serious API reductions upstream in https://github.com/kubernetes/cloud-provider-azure. As soon as these improvements stabilize, we'll integrate them into a future aks-engine release.

**Comment by andyzhangx** [comment]
if you got `vmssCSE time out` issue even after vmss update, you could use `reimage` vmss instance to mitigate, pls make sure no pods are running on that instance before reimage(run `kubectl cordon <agent-node-name>` before reimage)

- How to reimage vmss instance
![image](https://user-images.githubusercontent.com/4178417/67267233-738f4200-f4e4-11e9-964f-efe0147dc4a4.png)


```
Reimage virtual machine instance(s)
This action will reimage the selected instances, restoring them to their initial state. The instances will be restarted, and any local data will be lost. Do you want to reimage the selected instances?
```

 - root cause of `vmssCSE time out` issue after vmss update
after aks-engine `v0.40.0`, cluster-provision could not run twice, related PR: https://github.com/Azure/aks-engine/pull/1719, we should fix it in aks-engine  @jackfrancis 

cc @yangl900

**Comment by jackfrancis** [comment]
@andyzhangx thanks for that diagnosis, a fix to that has been included w/ v0.42.2, just published:

https://github.com/Azure/aks-engine/releases/tag/v0.42.2

**Comment by therock** [comment]
And do we run a aks-engine upgrade to get our vmss on the new version?

**Comment by jackfrancis** [comment]
Hi @therock, `aks-engine upgrade` would work. A more lightweight approach would be to use v0.42.2 of aks-engine and run `aks-engine scale` against the desired vmss-backed pool, giving it a n+1 count compared to the existing scaleset size. That would update the model for the scaleset, and then you could run "update model" against a single instance. I think that's safer than running upgrade, which will re-pave everything.

**Comment by ahmedspiir** [comment]
Hi @jac
When do i'll have it on azure? because i can see with the VMSs tag that i'm running version "aksEngineVersion : v0.41.4-aks".
Also is there a way to get the fix faster into my cluster?

**Comment by therock** [comment]
After running the aks-engine scale with v0.42.2 it fails with the same error 

> Provisioning of VM extension 'vmssCSE' has timed out. Extension installation may be taking too long, or extension status could not be obtained.

```
Error: Code="DeploymentFailed" Message="At least one resource deployment operation failed. Please list deployment operations for details. Please see https://aka.ms/arm-debug for usage details." Details=[{"code":"Conflict","message":"{
  "status": "Failed",
  "error": {
    "code": "ResourceDeploymentFailure",
    "message": "The resource operation completed with terminal provisioning state 'Failed'.",
    "details": [
      {
        "code": "VMExtensionProvisioningTimeout",
        "message": "Provisioning of VM extension 'vmssCSE' has timed out. Extension installation may be taking too long, or extension status could not be obtained."
      }
    ]
  }
}"}]
```

**Comment by acm-073** [comment]
> Hi @jackfrancis
> When do i'll have it on azure? because i can see with the VMSs tag that i'm running version "aksEngineVersion : v0.41.4-aks".
> Also is there a way to get the fix faster into my cluster?

Same question here - we're running into the very problems described here on a managed AKS cluster, so the question is when the fixes described here will be available there.

**Comment by therock** [comment]
After the failed `aks-engine scale` with `v0.42.2`
I did try `aks-engine upgrade` from version `1.16.0` to `1.16.2` and it failed at the last step ```Finished ARM Deployment``` with the same error 

> Provisioning of VM extension 'vmssCSE' has timed out. Extension installation may be taking too long, or extension status could not be obtained.

The upgraded master-0 node works fine even with that failed error so i ran it again 4 more times for the rest of the masters. Now all masters are upgraded to version 1.16.2 and it was suppose todo the `agent scale sets` but failed 20 secs later

```
INFO[0012] Starting upgrade of master nodes...
INFO[0012] masterNodesInCluster: 5
INFO[0012] Master VM: k8s-master-33419389-0 is upgraded to expected orchestrator version
INFO[0012] Master VM: k8s-master-33419389-1 is upgraded to expected orchestrator version
INFO[0012] Master VM: k8s-master-33419389-2 is upgraded to expected orchestrator version
INFO[0012] Master VM: k8s-master-33419389-3 is upgraded to expected orchestrator version
INFO[0012] Master VM: k8s-master-33419389-4 is upgraded to expected orchestrator version
INFO[0012] Expected master count: 5, Creating 0 more master VMs
INFO[0012] Deploying the agent scale sets ARM template...
INFO[0012] Starting ARM Deployment agentscaleset-19-10-23T12.06.49-1177339277 in resource group xxxxxx. This will take some time...
INFO[0040] Finished ARM Deployment (agentscaleset-19-10-23T12.06.49-1177339277). Error: Code="DeploymentFailed" Message="At least one resource deployment operation failed. Please list deployment operations for details. Please see https://aka.ms/arm-debug for usage details." Details=[{"code":"Conflict","message":"{
  \"status\": \"Failed\",
  \"error\": {
    \"code\": \"ResourceDeploymentFailure\",
    \"message\": \"The resource operation completed with terminal provisioning state 'Failed'.\",
    \"details\": [
      {
        \"code\": \"VMExtensionProvisioningTimeout\",
        \"message\": \"Provisioning of VM extension 'vmssCSE' has timed out. Extension installation may be taking too long, or extension status could not be obtained.\"
      }
    ]
  }
}"}]
ERRO[0040] error applying upgrade template in upgradeAgentScaleSets: Code="DeploymentFailed" Message="At least one resource deployment operation failed. Please list deployment operations for details. Please see https://aka.ms/arm-debug for usage details." Details=[{"code":"Conflict","message":"{
  \"status\": \"Failed\",
  \"error\": {
    \"code\": \"ResourceDeploymentFailure\",
    \"message\": \"The resource operation completed with terminal provisioning state 'Failed'.\",
    \"details\": [
      {
        \"code\": \"VMExtensionProvisioningTimeout\",
        \"message\": \"Provisioning of VM extension 'vmssCSE' has timed out. Extension installation may be taking too long, or extension status could not be obtained.\"
      }
    ]
  }
}"}]
INFO[0040] Resuming cluster autoscaler, replica count: 1
Error: upgrading cluster: Code="DeploymentFailed" Message="At least one resource deployment operation failed. Please list deployment operations for details. Please see https://aka.ms/arm-debug for usage details." Details=[{"code":"Conflict","message":"{
  \"status\": \"Failed\",
  \"error\": {
    \"code\": \"ResourceDeploymentFailure\",
    \"message\": \"The resource operation completed with terminal provisioning state 'Failed'.\",
    \"details\": [
      {
        \"code\": \"VMExtensionProvisioningTimeout\",
        \"message\": \"Provisioning of VM extension 'vmssCSE' has timed out. Extension installation may be taking too long, or extension status could not be obtained.\"
      }
    ]
  }
}"}]
```

**Comment by zachomedia** [comment]
@jackfrancis We upgraded to 1.15.5 with aks-engine 0.42.2 and enabled the v2 backoff. But it seems that when it hits a rate limit scenario it just hammers the Azure API and the only way to recover is to turn off controller-manager for a while to clear it

For example:

```
The server rejected the request because too many requests have been received for this subscription. (Code: OperationNotAllowed) {"operationgroup":"HighCostGetVMScaleSet30Min","starttime":"2019-10-23T14:18:11.960853+00:00","endtime":"2019-10-23T14:33:11.960853+00:00","allowedrequestcount":900,"measuredrequestcount":3157} (Code: TooManyRequests, Target: HighCostGetVMScaleSet30Min)
```

(We also hit the HighCostGetVMScaleSet3Min one too)

**Comment by AustinSmart** [comment]
Will the fixes being addressed over at the [AKS issue](https://github.com/Azure/AKS/issues/1278#issuecomment-545234688) be applicable here?

**Comment by sylus** [comment]
Yeah as @zachomedia said it appears worse and we barely get 15-20 mins of proper Kubernetes operation before rate limited again. @devigned @jackfrancis can we arbitratrily increase our limit?

Our clients are getting pretty angry at the state of things this week and is putting a lot of pressure on us. Really worried will result in us having to move workloads somewhere else which really don't want. Have been stalled for the past 2-3 days. We can maybe increase our support ticket to priority.

**Comment by jackfrancis** [comment]
For folks who are having issues w/ VMSS CSE timeouts (not necessarily related to throttling), there has been an identified CSE bug being triaged. This bug correlates with folks experiencing this issue last week. (This CSE bug has nothing to do w/ AKS Engine's CSE script(s).)

If you have one or more VMSS instances in this state, please try manually re-imaging the instance. We've seen that workaround help restore nodes for folks.

And please report back if that unblocks you.

And apologies. :(

**Comment by zachomedia** [comment]
Ok, so I managed to get our cluster back into a good state for now. Seems that running most operations that cause a disk unmount can trigger the problem again (@sylus can fill in more there)

Basically to find out which disk is stuck, I looked into the controller-manager logs and saw a bunch of:

```
I1023 20:02:29.782144       1 attacher.go:89] Attach volume "/subscriptions/$SUBSCRIPTION/resourceGroups/k8s-cancentral-01-dev-rg/providers/Microsoft.Compute/disks/k8s-cancentral-01-development-dyna-pvc-$PVC_NAME" to instance "k8s-linuxpool1-12345678-vmss00000c" failed with disk(/subscriptions/$SUBSCRIPTION/resourceGroups/k8s-cancentral-01-dev-rg/providers/Microsoft.Compute/disks/k8s-cancentral-01-development-dyna-pvc-$PVC_NAME) already attached to node(/subscriptions/$SUBSCRIPTION/resourceGroups/k8s-cancentral-01-dev-rg/providers/Microsoft.Compute/virtualMachineScaleSets/k8s-linuxpool1-12345678-vmss/virtualMachines/k8s-linuxpool1-12345678-vmss_10), could not be attached to node(k8s-linuxpool1-12345678-vmss00000c)
I1023 20:02:29.782187       1 actual_state_of_world.go:322] Volume "kubernetes.io/azure-disk//subscriptions/$SUBSCRIPTION/resourceGroups/k8s-cancentral-01-dev-rg/providers/Microsoft.Compute/disks/k8s-cancentral-01-development-dyna-pvc-$PVC_NAME" is already added to attachedVolume list to node "k8s-linuxpool1-12345678-vmss_10", update device path ""
E1023 20:02:29.782390       1 nestedpendingoperations.go:270] Operation for "\"kubernetes.io/azure-disk//subscriptions/$SUBSCRIPTION/resourceGroups/k8s-cancentral-01-dev-rg/providers/Microsoft.Compute/disks/k8s-cancentral-01-development-dyna-pvc-$PVC_NAME\"" failed. No retries permitted until 2019-10-23 20:02:30.282260017 +0000 UTC m=+391.358857249 (durationBeforeRetry 500ms). Error: "AttachVolume.Attach failed for volume \"pvc-$PVC_NAME\" (UniqueName: \"kubernetes.io/azure-disk//subscriptions/$SUBSCRIPTION/resourceGroups/k8s-cancentral-01-dev-rg/providers/Microsoft.Compute/disks/k8s-cancentral-01-development-dyna-pvc-$PVC_NAME\") from node \"k8s-linuxpool1-12345678-vmss00000c\" : disk(/subscriptions/$SUBSCRIPTION/resourceGroups/k8s-cancentral-01-dev-rg/providers/Microsoft.Compute/disks/k8s-cancentral-01-development-dyna-pvc-$PVC_NAME) already attached to node(/subscriptions/$SUBSCRIPTION/resourceGroups/k8s-cancentral-01-dev-rg/providers/Microsoft.Compute/virtualMachineScaleSets/k8s-linuxpool1-12345678-vmss/virtualMachines/k8s-linuxpool1-12345678-vmss_10), could not be attached to node(k8s-linuxpool1-12345678-vmss00000c)"
I1023 20:02:29.782447       1 event.go:258] Event(v1.ObjectReference{Kind:"Pod", Namespace:"adp", Name:"appname-7587887f5d-8pndl", UID:"2b777c28-859e-45cd-984f-af2734a436a5", APIVersion:"v1", ResourceVersion:"10305952", FieldPath:""}): type: 'Warning' reason: 'FailedAttachVolume' AttachVolume.Attach failed for volume "pvc-$PVC_NAME" : disk(/subscriptions/$SUBSCRIPTION/resourceGroups/k8s-cancentral-01-dev-rg/providers/Microsoft.Compute/disks/k8s-cancentral-01-development-dyna-pvc-$PVC_NAME) already attached to node(/subscriptions/$SUBSCRIPTION/resourceGroups/k8s-cancentral-01-dev-rg/providers/Microsoft.Compute/virtualMachineScaleSets/k8s-linuxpool1-12345678-vmss/virtualMachines/k8s-linuxpool1-12345678-vmss_10), could not be
attached to node(k8s-linuxpool1-12345678-vmss00000c)
```

To fix it, I ran `az vmss list-instances --resource-group k8s-cancentral-01-dev-rg --name k8s-linuxpool1-12345678-vmss --query '[].[name, storageProfile.dataDisks[]]' | less`, found the disk and got the lun number. I then ran `az vmss disk detach --resource-group k8s-cancentral-01-dev-rg --vmss-name k8s-linuxpool1-12345678-vmss --instance-id 12 --lun 3` to detach it.

Once it was detached, the cluster slowly started to recover.

**Comment by sylus** [comment]
Re: Above this does seem to be the problem. I was able to reproduce a base case by just deleting a few pods and / or running a helm upgrade deployment and then this started to trigger the following errors related to a disk unmount almost right away.

```sh
I1023 23:49:18.644247       1 pv_controller.go:1270] isVolumeReleased[pvc-b2220d20-9e95-4973-923d-95cc6e49ff4c]: volume is released
I1023 23:49:18.644258       1 pv_controller.go:1270] isVolumeReleased[pvc-653fdf5a-7408-460e-89d3-fdd0a6dd5fdf]: volume is released
E1023 23:49:19.821907       1 goroutinemap.go:150] Operation for "delete-pvc-b2220d20-9e95-4973-923d-95cc6e49ff4c[52a16d3e-b5dd-4cc1-a64e-b03f6d61948b]" failed. No retries permitted until 2019-10-23 23:51:21.821868366 +0000 UTC m=+12222.629161185 (durationBeforeRetry 2m2s). Error: "compute.DisksClient#Delete: Failure sending request: StatusCode=0 -- Original Error: autorest/azure: Service returned an error. Status=<nil> Code=\"OperationNotAllowed\" Message=\"Disk k8s-cancentral-01-development-dyna-pvc-b2220d20-9e95-4973-923d-95cc6e49ff4c is attached to VM /subscriptions/$SUBSCRIPTION/resourceGroups/k8s-cancentral-01-dev-rg/providers/Microsoft.Compute/virtualMachineScaleSets/k8s-linuxpool1-28391316-vmss/virtualMachines/k8s-linuxpool1-28391316-vmss_12.\""
E1023 23:49:19.826714       1 goroutinemap.go:150] Operation for "delete-pvc-653fdf5a-7408-460e-89d3-fdd0a6dd5fdf[4c641582-1171-4c98-8189-29185623fc1c]" failed. No retries permitted until 2019-10-23 23:51:21.826677075 +0000 UTC m=+12222.633969894 (durationBeforeRetry 2m2s). Error: "compute.DisksClient#Delete: Failure sending request: StatusCode=0 -- Original Error: autorest/azure: Service returned an error. Status=<nil> Code=\"OperationNotAllowed\" Message=\"Disk k8s-cancentral-01-development-dyna-pvc-653fdf5a-7408-460e-89d3-fdd0a6dd5fdf is attached to VM /subscriptions/$SUBSCRIPTION/resourceGroups/k8s-cancentral-01-dev-rg/providers/Microsoft.Compute/virtualMachineScaleSets/k8s-linuxpool1-28391316-vmss/virtualMachines/k8s-linuxpool1-28391316-vmss_12.\""
```

The suspicion is as we have multiple teams (re)-deploying their apps etc that enough of these disk failure issues eventually make us hit the rate limits set by Azure. Therefore other operations against the VMSS don't succeed once it gets into this state. Then as mentioned above we need to stop the controller-manager pod for a while to clear the rate limit as illustrated below.

```sh
The server rejected the request because too many requests have been received for this subscription. (Code: OperationNotAllowed) {"operationgroup":"HighCostGetVMScaleSet30Min","starttime":"2019-10-23T14:18:11.960853+00:00","endtime":"2019-10-23T14:33:11.960853+00:00","allowedrequestcount":900,"measuredrequestcount":3157} (Code: TooManyRequests, Target: HighCostGetVMScaleSet30Min)
```

In the controller-manager it will list the PVC and instance ID of the disk that can't be unattached and we can use the PVC number to find the LUN below.

```sh
az vmss list-instances --resource-group k8s-cancentral-01-dev-rg --name k8s-linuxpool1-12345678-vmss --query '[].[name, storageProfile.dataDisks[]]' 
```

We then for all disks listed in controller manager that have this problem need to run the following.

```sh
az vmss disk detach --resource-group k8s-cancentral-01-dev-rg --vmss-name k8s-linuxpool1-12345678-vmss --instance-id $ID --lun $LUN

**Comment by jackfrancis** [comment]
@sylus @zachomedia in your failure scenarios, are you ever encountering this error:

```
Cannot attach data disk '<disk_id>' to VM '<vmss_id>' because the disk is currently being detached or the last detach operation failed. Please wait until the disk is completely detached and then try again or delete/detach the disk explicitly again."
```

And if so, are you observing that the disk it's complaining about seems, in fact, to be totally unattached?

We encountered this with another customer, and were able to follow your guidance to manually detach the offending disk (even though it wasn't attached to anything! — we detached it from the vmss instance id that it was trying to attach itself to 🤷‍♂ ).

In any event, FYI for folks continuing to struggle with this.

**Comment by zachomedia** [comment]
@jackfrancis Yeah, we've seen that error in our logs too. I don't think we've ever checked if it was actually attached or not, usually we just detach it through the cli.

We did have a weird state today where apparently one of our instances had a disk attached that no longer existed so all other disk operations failed. Once we removed that attachment, it recovered.

**Comment by jackfrancis** [comment]
@zachomedia How did you get the lun number in the case where the disk is not actually attached to any VMSS instances?

In our troubleshooting the following command didn't yield the lun in such a scenario:

```
az vmss list-instances --resource-group <rg> --name <vmss_name> --query '[].[name, storageProfile.dataDisks[]]'
```

**Comment by zachomedia** [comment]
@jackfrancis Oh, I see, for all of our cases the disk was in the list. So I guess that means it was attached.

**Comment by jackfrancis** [comment]
@sylus @zachomedia do you think this is an appropriate repro?:

1) Install this statefulset on a cluster w/ 50 VMSS nodes:

```
$ cat statefulset.yaml 
apiVersion: apps/v1
kind: StatefulSet
metadata:
  name: web
spec:
  selector:
    matchLabels:
      app: nginx
  serviceName: "nginx"
  podManagementPolicy: "Parallel"
  replicas: 50
  template:
    metadata:
      labels:
        app: nginx
    spec:
      terminationGracePeriodSeconds: 10
      containers:
      - name: nginx
        image: k8s.gcr.io/nginx-slim:0.8
        ports:
        - containerPort: 80
          name: web
        volumeMounts:
        - name: www
          mountPath: /usr/share/nginx/html
  volumeClaimTemplates:
  - metadata:
      name: www
    spec:
      accessModes: [ "ReadWriteOnce" ]
      storageClassName: "managed-standard"
      resources:
        requests:
          storage: 1Gi
```

(note the ratio of nodes-replicas is 1:1)

2) When pods come online, start deleting them
3) Statefulset reconciliation will spin up more pods to fulfill replica count

I wonder if the above will induce the weird zombie detach state we're seeing.

**Comment by zachomedia** [comment]
@jackfrancis 

I would say that you should have a pretty reasonable chance at reproducing the issue with that setup. Our cluster is much smaller (about 5 nodes) and usually just a couple of pods with disks being deleted can trigger it.

**Comment by jackfrancis** [comment]
@zachomedia Node count is static? I.e., disk re-attachment operations aren't happening as a result of underlying VMSS instances disappearing, re-appearing, etc?

**Comment by zachomedia** [comment]
@jackfrancis That's correct, node count is static.

**Comment by jackfrancis** [comment]
(So far unable to repro, but will keep trying.)

It's also possible that disk detach/attach operations during throttle events are the edge case causing this behavior (my test cluster is not being actively throttled atm).

**Comment by zachomedia** [comment]
@jackfrancis So something you can try.. Seems most of our problems now stem from PVCs being deleted (one of our teams deletes their deployments and re-recreates it right now). We seem to get two things:

1. A disk is still attached to a VMSS instance but apparently no longer exists. No disk operations succeed until we detach that disk (the error appears when we attempt to manually detach a different disk that was failing - sorry, I forgot to save the message)
2. Some deletes seem to never unmount the disk (maybe it passed some timeouts and gave up trying to unmount). See a bunch of: 

```
I1025 15:21:21.911563       1 pv_controller.go:1270] isVolumeReleased[pvc-PVC1]: volume is released
I1025 15:21:21.911914       1 pv_controller.go:1270] isVolumeReleased[pvc-PVC2]: volume is released
I1025 15:21:21.913827       1 pv_controller.go:1270] isVolumeReleased[pvc-PVC3]: volume is released
E1025 15:21:23.028284       1 goroutinemap.go:150] Operation for "delete-pvc-PVC2[5dcfd7e1-e61b-4206-9cac-6e7758e3366c]" failed. No retries permitted until 2019-10-25 15:23:25.028194036 +0000 UTC m=+154545.835486755 (durationBeforeRetry 2m2s). Error: "compute.DisksClient#Delete: Failure sending request: StatusCode=0 -- Original Error: autorest/azure: Service returned an error. Status=<nil> Code=\"OperationNotAllowed\" Message=\"Disk k8s-cancentral-01-development-dyna-pvc-PVC2 is attached to VM /subscriptions/<subscription>/resourceGroups/k8s-cancentral-01-dev-rg/providers/Microsoft.Compute/virtualMachineScaleSets/k8s-linuxpool1-12345678-vmss/virtualMachines/k8s-linuxpool1-12345678-vmss_12.\""
E1025 15:21:23.029986       1 goroutinemap.go:150] Operation for "delete-pvc-PVC3[3507cbda-868e-4198-b771-620234d258b5]" failed. No retries permitted until 2019-10-25 15:23:25.029908242 +0000 UTC m=+154545.837201061 (durationBeforeRetry 2m2s). Error: "compute.DisksClient#Delete: Failure sending request: StatusCode=0 -- Original Error: autorest/azure: Service returned an error. Status=<nil> Code=\"OperationNotAllowed\" Message=\"Disk k8s-cancentral-01-development-dyna-pvc-PVC3 is attached to VM /subscriptions/<subscription>/resourceGroups/k8s-cancentral-01-dev-rg/providers/Microsoft.Compute/virtualMachineScaleSets/k8s-linuxpool1-12345678-vmss/virtualMachines/k8s-linuxpool1-12345678-vmss_12.\""
E1025 15:21:23.069595       1 goroutinemap.go:150] Operation for "delete-pvc-PVC1[4d9dbfa3-1dba-4478-9eae-f475d033257a]" failed. No retries permitted until 2019-10-25 15:23:25.069557061 +0000 UTC m=+154545.876849880 (durationBeforeRetry 2m2s). Error: "compute.DisksClient#Delete: Failure sending request: StatusCode=0 -- Original Error: autorest/azure: Service returned an error. Status=<nil> Code=\"OperationNotAllowed\" Message=\"Disk k8s-cancentral-01-development-dyna-pvc-PVC1 is attached to VM /subscriptions/<subscription>/resourceGroups/k8s-cancentral-01-dev-rg/providers/Microsoft.Compute/virtualMachineScaleSets/k8s-linuxpool1-12345678-vmss/virtualMachines/k8s-linuxpool1-12345678-vmss_12.\""

```

**Comment by andyzhangx** [comment]
@zachomedia You need to drain `k8s-linuxpool1-12345678-vmss_12` and then reimage `k8s-linuxpool1-12345678-vmss_12` to clean that vmss instance state as a workaround.

**Comment by sylus** [comment]
Problems again today with 5 of our 6 VM's in a failed state and can't even reimage as get this error. I also did launch a few AKS clusters over the weekend and as soon as turned off over night, they all are in failed state with disk issues. Really hoping a fix is forthcoming, this is plainly reproducible.

```sh

**Comment by sylus** [comment]
Failed to reimage virtual machine instances k8s-linuxpool1-12345678-vmss_12, k8s-linuxpool1-12345678-vmss_10, k8s-linuxpool1-12345678-vmss_9, k8s-linuxpool1-12345678-vmss_11. Error: The processing of VM 'k8s-linuxpool1-12345678-vmss_10' is halted because of one or more disk processing errors encountered by VM 'k8s-linuxpool1-12345678-vmss_12' in the same Availability Set. Please resolve the error with VM 'k8s-linuxpool1-12345678-vmss_12' before retrying the operation.
```

**Comment by jackfrancis** [comment]
@sylus is the StatefulSet spec here not a viable repro input for inducing this symptom on a test cluster?

https://github.com/Azure/aks-engine/issues/2221

As the issue describes I *was* able to witness some badness (described in the issue w/ the working remediation steps I came up with at the time), but I haven't been able to reliably repeat all the badness so many folks are seeing now. Would love to get that repro process so that we can more effectively help drive fixes.

Thanks for hanging in there. :/

**Comment by craiglpeters** [comment]
A VMSS bug in the update on Oct 17 was identified and remediated globally over Oct 28 and 29. Disks should no longer be stuck in the 'detaching' state in VMSS, and so any Kubernetes operations should now be able to proceed without running into this issue. If you observe any new instances of this same problem please reopen this bug and I'll work to determine the cause.

**Comment by mbarry-msdn** [comment]
We're getting error with version 1.22.6: "MessageProvisioning of VM extension vmssCSE has timed out. Extension provisioning has taken too long to complete. The extension last reported "Plugin enabled". More information on troubleshooting is available at https://aka.ms/VMExtensionCSELinuxTroubleshootTimeMonday,"

**Comment by Lu1234** [comment]
> We're getting error with version 1.22.6: "MessageProvisioning of VM extension vmssCSE has timed out. Extension provisioning has taken too long to complete. The extension last reported "Plugin enabled". More information on troubleshooting is available at https://aka.ms/VMExtensionCSELinuxTroubleshootTimeMonday,"

I'm facing the same error while nothing being changed to the cluster,really need some advice with this.

**Comment by CecileRobertMichon** [comment]
Hi all, I suggest opening a new issue in https://github.com/Azure/AKS/issues with details of the problem/error you are facing.

I want to make sure you're getting the help you need. This is a closed issue from 3 years ago in a deprecated project (https://github.com/Azure/aks-engine#project-status) so commenting on here likely won't get the right people to look into it.
